From 196f8751be31612ff0b1aab42d2f4fabde09ba03 Mon Sep 17 00:00:00 2001
From: Etienne Carriere <etienne.carriere@foss.st.com>
Date: Fri, 12 Jan 2024 12:43:46 +0100
Subject: [PATCH 0432/1695] drivers: regulator: DT property
 regulator-enable-ramp-delay

Implement regulator DT property regulator-enable-ramp-delay.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git c6f216ae52c8c30927ac63b005ad2cd136e77586 drivers: regulator: DT property regulator-enable-ramp-delay]
Change-Id: I1a392dc92bd9227bf41aad06ccc54ff81afee51a
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
---
 core/drivers/regulator/regulator.c    | 2 ++
 core/drivers/regulator/regulator_dt.c | 8 ++++++++
 core/include/drivers/regulator.h      | 2 ++
 3 files changed, 12 insertions(+)

diff --git a/core/drivers/regulator/regulator.c b/core/drivers/regulator/regulator.c
index c6b34d66e..b7b9f1bf0 100644
--- a/core/drivers/regulator/regulator.c
+++ b/core/drivers/regulator/regulator.c
@@ -90,6 +90,8 @@ static TEE_Result regulator_refcnt_enable(struct regulator *regulator)
 
 			return res;
 		}
+
+		udelay(regulator->enable_ramp_delay_us);
 	}
 
 	regulator->refcount++;
diff --git a/core/drivers/regulator/regulator_dt.c b/core/drivers/regulator/regulator_dt.c
index 196642216..c1ff6dba6 100644
--- a/core/drivers/regulator/regulator_dt.c
+++ b/core/drivers/regulator/regulator_dt.c
@@ -327,6 +327,14 @@ static TEE_Result parse_dt(const void *fdt, int node,
 		     regulator->ramp_delay_uv_per_us);
 	}
 
+	cuint = fdt_getprop(fdt, node, "regulator-enable-ramp-delay", NULL);
+	if (cuint) {
+		regulator->enable_ramp_delay_us = fdt32_to_cpu(*cuint);
+		FMSG("%s: enable ramp delay = %u (us)",
+		     regulator_name(regulator),
+		     regulator->enable_ramp_delay_us);
+	}
+
 	return TEE_SUCCESS;
 }
 
diff --git a/core/include/drivers/regulator.h b/core/include/drivers/regulator.h
index a991ae8b4..de162b636 100644
--- a/core/include/drivers/regulator.h
+++ b/core/include/drivers/regulator.h
@@ -94,6 +94,7 @@ struct regulator_voltages_desc {
  * @min_uv: Min possible voltage level in microvolt (uV)
  * @max_uv: Max possible voltage level in microvolt (uV)
  * @ramp_delay_uv_per_us: Voltage level change delay in uV/s
+ * @enable_ramp_delay_us: Delay after enable, in microseconds (us)
  * @cur_uv: Current voltage level in microvolt (uV)
  * @flags: REGULATOR_* property flags
  * @refcount: Regulator enable request reference counter
@@ -110,6 +111,7 @@ struct regulator {
 	int min_uv;
 	int max_uv;
 	unsigned int ramp_delay_uv_per_us;
+	unsigned int enable_ramp_delay_us;
 	/* Fields internal to regulator framework */
 	int cur_uv;
 	unsigned int flags;
-- 
2.47.2

